File: service-rest/src/test/java/com/bnorm/barkeep/rest/AbstractRestServiceTest.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest;

import java.util.Properties;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;

import org.assertj.core.api.Condition;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.runner.RunWith;
import org.springframework.boot.context.embedded.LocalServerPort;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.util.EnvironmentTestUtils;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;
import org.testcontainers.containers.BindMode;
import org.testcontainers.containers.GenericContainer;

import com.bnorm.barkeep.Application;
import com.bnorm.barkeep.db.DbUserService;
import com.bnorm.barkeep.model.Bar;
import com.bnorm.barkeep.model.BarValue;
import com.bnorm.barkeep.model.Book;
import com.bnorm.barkeep.model.BookValue;
import com.bnorm.barkeep.model.Component;
import com.bnorm.barkeep.model.ComponentValue;
import com.bnorm.barkeep.model.HasId;
import com.bnorm.barkeep.model.Ingredient;
import com.bnorm.barkeep.model.IngredientValue;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.model.RecipeValue;
import com.bnorm.barkeep.model.User;
import com.bnorm.barkeep.model.UserValue;
import com.squareup.moshi.Moshi;

import okhttp3.Authenticator;
import okhttp3.Credentials;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ContextConfiguration(classes = Application.class, initializers = AbstractRestServiceTest.Initializer.class)
public abstract class AbstractRestServiceTest {

  private static final int MYSQL_PORT = 3306;

  @ClassRule
  public static final GenericContainer mysql;

  static {
    mysql = new GenericContainer("mysql:latest");
    mysql.withClasspathResourceMapping("com/bnorm/barkeep/db/init", "/docker-entrypoint-initdb.d", BindMode.READ_ONLY);
    mysql.addEnv("MYSQL_ROOT_PASSWORD", "root");
    mysql.addExposedPort(MYSQL_PORT);
  }


  public static class Initializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

    @Override
    public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
      EnvironmentTestUtils.addEnvironment("testcontainers",
                                          configurableApplicationContext.getEnvironment(),
                                          "barkeep.db.host=" + mysql.getContainerIpAddress(),
                                          "barkeep.db.port=" + mysql.getMappedPort(MYSQL_PORT));
    }
  }


  // constants
  static final int CODE_SUCCESS = 200;
  static final int BAD_REQUEST = 400;
  static final int CODE_UNAUTHORIZED = 401;
  static final int CODE_NOT_FOUND = 404;

  static final User TEST_USER = UserValue.builder()
                                         .setUsername("joe")
                                         .setPassword("testmore")
                                         .setEmail("joe@test.more")
                                         .build();

  // conditions
  static final Condition<? super HasId> VALID_ID = new Condition<>(b -> b.getId() != null && b.getId() != -1,
                                                                   "valid id");

  // test service

  @LocalServerPort
  private int port;

  BarkeepService service;

  @BeforeClass
  public static void setupDatabase() {
    // todo figure out a better way to create test user
    Properties properties = new Properties();
    properties.put("hibernate.connection.url",
                   String.format("jdbc:mysql://%s:%d/barkeep",
                                 mysql.getContainerIpAddress(),
                                 mysql.getMappedPort(MYSQL_PORT)));
    properties.put("hibernate.connection.username", "root");
    properties.put("hibernate.connection.password", "root");

    EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("com.bnorm.barkeep.jpa",
                                                                                       properties);
    EntityManager entityManager = entityManagerFactory.createEntityManager();
    DbUserService userService = new DbUserService(entityManager);

    BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
    userService.createUser(UserValue.builder()
                                    .setUsername(TEST_USER.getUsername())
                                    .setPassword(passwordEncoder.encode(TEST_USER.getPassword()))
                                    .setEmail(TEST_USER.getEmail())
                                    .build());
  }

  @Before
  public void setup() throws Exception {
    HttpUrl.Builder urlBuilder = new HttpUrl.Builder();
    urlBuilder.scheme("http");
    urlBuilder.host("localhost");
    urlBuilder.port(port);
    //    urlBuilder.addPathSegment("api");
    //    urlBuilder.addPathSegment("rest");
    //    urlBuilder.addPathSegment("v1");
    urlBuilder.addPathSegment("barkeep");
    // This last, empty segment adds a trailing '/' which is required for relative paths in the annotations
    urlBuilder.addPathSegment("");
    HttpUrl url = urlBuilder.build();


    Authenticator authenticator = (route, response) -> {
      return response.request()
                     .newBuilder()
                     .header("Authorization", Credentials.basic(TEST_USER.getUsername(), TEST_USER.getPassword()))
                     .build();
    };
    OkHttpClient client = new OkHttpClient.Builder().addInterceptor(new CookieInterceptor(url))
                                                    // .addInterceptor(new WireTraceInterceptor())
                                                    .authenticator(authenticator).build();
    Moshi moshi = new Moshi.Builder().add((type, annotations, m) -> {
      if (type.equals(Recipe.class)) {
        return RecipeValue.jsonAdapter(m);
      } else if (type.equals(Bar.class)) {
        return BarValue.jsonAdapter(m);
      } else if (type.equals(Book.class)) {
        return BookValue.jsonAdapter(m);
      } else if (type.equals(Ingredient.class)) {
        return IngredientValue.jsonAdapter(m);
      } else if (type.equals(Component.class)) {
        return ComponentValue.jsonAdapter(m);
      } else if (type.equals(User.class)) {
        return UserValue.jsonAdapter(m);
      } else {
        return null;
      }
    }).add(ImmutableSetAdapter.FACTORY).build();
    Retrofit retrofit = new Retrofit.Builder().baseUrl(url)
                                              .client(client)
                                              .addConverterFactory(MoshiConverterFactory.create(moshi))
                                              .build();

    service = retrofit.create(BarkeepService.class);
  }
}

File: service-rest/src/test/kotlin/com/bnorm/barkeep/rest/AbstractRestServiceTest.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest

import com.bnorm.barkeep.RestApplication
import com.bnorm.barkeep.db.DbUserService
import com.bnorm.barkeep.model.Bar
import com.bnorm.barkeep.model.BarValue
import com.bnorm.barkeep.model.Book
import com.bnorm.barkeep.model.BookValue
import com.bnorm.barkeep.model.Component
import com.bnorm.barkeep.model.ComponentValue
import com.bnorm.barkeep.model.Ingredient
import com.bnorm.barkeep.model.IngredientValue
import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.model.RecipeValue
import com.bnorm.barkeep.model.User
import com.bnorm.barkeep.model.UserValue
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import io.kotlintest.matchers.Matcher
import io.kotlintest.matchers.Matchers
import okhttp3.Authenticator
import okhttp3.Credentials
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import org.junit.Before
import org.junit.BeforeClass
import org.junit.ClassRule
import org.junit.Rule
import org.springframework.boot.context.embedded.LocalServerPort
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.util.EnvironmentTestUtils
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.rules.SpringClassRule
import org.springframework.test.context.junit4.rules.SpringMethodRule
import org.testcontainers.containers.BindMode
import org.testcontainers.containers.GenericContainer
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import java.util.Properties
import javax.persistence.Persistence

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ContextConfiguration(classes = arrayOf(RestApplication::class),
                      initializers = arrayOf(AbstractRestServiceTest.Initializer::class))
@DirtiesContext // forces a reload of app for every test - required because database is reloaded
abstract class AbstractRestServiceTest : Matchers {
  companion object {
    val beValid = object : Matcher<Long> {
      override fun test(value: Long) {
        if (value == -1L)
          throw AssertionError("$value is not a valid ID")
      }
    }

    @get:ClassRule
    @JvmStatic
    val SPRING_CLASS_RULE = SpringClassRule()

    private val MYSQL_PORT = 3306

    @get:ClassRule
    @JvmStatic
    val mysql: DockerContainer = DockerContainer("mysql:latest").apply {
      withClasspathResourceMapping("com/bnorm/barkeep/db/init", "/docker-entrypoint-initdb.d", BindMode.READ_ONLY)
      addEnv("MYSQL_ROOT_PASSWORD", "root")
      addExposedPort(MYSQL_PORT)
    }

    // constants
    internal val CODE_SUCCESS = 200
    internal val BAD_REQUEST = 400
    internal val CODE_UNAUTHORIZED = 401
    internal val CODE_NOT_FOUND = 404

    internal val TEST_USER: User = UserValue.builder()
            .setUsername("joe")
            .setPassword("testmore")
            .setEmail("joe@test.more")
            .build()

    @BeforeClass
    @JvmStatic
    fun setupDatabase() {
      // todo figure out a better way to create test user
      val address = mysql.getContainerIpAddress()
      val port = mysql.getMappedPort(MYSQL_PORT)

      val properties = Properties()
      properties.put("hibernate.connection.url", "jdbc:mysql://$address:$port/barkeep")
      properties.put("hibernate.connection.username", "root")
      properties.put("hibernate.connection.password", "root")

      val factory = Persistence.createEntityManagerFactory("com.bnorm.barkeep.jpa", properties)
      val em = factory.createEntityManager()
      val userService = DbUserService(em)

      val passwordEncoder = BCryptPasswordEncoder()
      userService.createUser(UserValue.builder()
                                     .setUsername(TEST_USER.username)
                                     .setPassword(passwordEncoder.encode(TEST_USER.password!!))
                                     .setEmail(TEST_USER.email)
                                     .build())
    }
  }

  @Rule
  val springMethodRule = SpringMethodRule()

  @LocalServerPort
  private val port: Int = 0


  class Initializer : ApplicationContextInitializer<ConfigurableApplicationContext> {
    override fun initialize(configurableApplicationContext: ConfigurableApplicationContext) {
      val address = mysql.getContainerIpAddress()
      val port = mysql.getMappedPort(MYSQL_PORT)
      EnvironmentTestUtils.addEnvironment("testcontainers",
                                          configurableApplicationContext.environment,
                                          "barkeep.db.host=$address",
                                          "barkeep.db.port=$port")
    }
  }

  // test service

  lateinit var service: BarkeepService

  @Before
  @Throws(Exception::class)
  fun setup() {
    val urlBuilder = HttpUrl.Builder()
    urlBuilder.scheme("http")
    urlBuilder.host("localhost")
    urlBuilder.port(port)
    urlBuilder.addPathSegment("rest")
    urlBuilder.addPathSegment("api")
    urlBuilder.addPathSegment("v1")
    // This last, empty segment adds a trailing '/' which is required for relative paths in the annotations
    urlBuilder.addPathSegment("")
    val url = urlBuilder.build()


    val authenticator = Authenticator { route, response ->
      response.request()
              .newBuilder()
              .header("Authorization", Credentials.basic(TEST_USER.username, TEST_USER.password))
              .build()
    }
    val client = OkHttpClient.Builder().addInterceptor(CookieInterceptor(url))
            .addInterceptor(WireTraceInterceptor())
            .authenticator(authenticator)
            .build()
    val moshi = Moshi.Builder().add(JsonAdapter.Factory { type, annotations, m ->
      if (type == Recipe::class.java) {
        return@Factory RecipeValue.jsonAdapter(m)
      } else if (type == Bar::class.java) {
        return@Factory BarValue.jsonAdapter(m)
      } else if (type == Book::class.java) {
        return@Factory BookValue.jsonAdapter(m)
      } else if (type == Ingredient::class.java) {
        return@Factory IngredientValue.jsonAdapter(m)
      } else if (type == Component::class.java) {
        return@Factory ComponentValue.jsonAdapter(m)
      } else if (type == User::class.java) {
        return@Factory UserValue.jsonAdapter(m)
      } else {
        return@Factory null
      }
    }).add(ImmutableSetAdapter.FACTORY).build()
    val retrofit = Retrofit.Builder().baseUrl(url)
            .client(client)
            .addConverterFactory(MoshiConverterFactory.create(moshi))
            .build()

    service = retrofit.create(BarkeepService::class.java)
  }
}

class DockerContainer(image: String) : GenericContainer<DockerContainer>(image)
